packages/bsky/src/notifications.ts

```diff
@@ -0,0 +1,253 @@
+      reason,
+      reasonSubject: subjectUri, // if like/reply/quote/emtion, the post which was liked/replied to/mention is in/or quoted. if custom feed liked, the feed which was liked
+      recordUri,
+    } = notif
+    // 1. Get author's display name
+    const displayName = await this.db.db
+      .selectFrom('profile')
+      .where('creator', '=', authorDid)
+      .select(['displayName'])
+      .executeTakeFirst()
+
+    // if no display name, dont send notification
+    if (!displayName || !displayName?.displayName) {
+      return
+    }
+    const author = displayName.displayName
+
+    // 2. Get post data content
+    // if reply, quote, or mention, get URI of the postRecord
+    // if like, or custom feed like, or repost get the URI of the reasonSubject
+    // if follow, get the URI of the author's profile
+    let title = ''
+    let body = ''
+
+    // check follow first and mention first because they don't have subjectUri and return
+    if (reason === 'follow') {
+      title = 'New follower!'
+      body = `${author} has followed you`
+      return { title, body }
+    } else if (reason === 'mention') {
+      title = `${author} mentioned you`
+      const mentionedPostData = await this.db.db
+        .selectFrom('post')
+        .where('uri', '=', recordUri)
+        .selectAll()
+        .executeTakeFirst()
+      body = mentionedPostData?.text || ''
+      return { title, body }
+    }
+
+    // if no subjectUri, don't send notification
+    // at this point, subjectUri should exist for all the other reasons
+    if (!subjectUri) {
+      return
+    }
+
+    // if no post data, don't send notification
+    const postData = await this.db.db
+      .selectFrom('post')
+      .where('uri', '=', subjectUri)
+      .selectAll()
+      .executeTakeFirst()
+    if (!postData) {
+      return
+    }
+
+    if (reason === 'like') {
+      title = `${author} liked your post`
+      body = postData?.text || ''
+      // custom feed like
+      if (subjectUri?.includes('feed.generator')) {
+        title = `${author} liked your custom feed`
+        body = `${new AtUri(subjectUri).rkey}`
+      }
+    } else if (reason === 'reply') {
+      title = `${author} replied to your post`
+      body = postData?.text || ''
+    } else if (reason === 'quote') {
+      title = `${author} quoted your post`
+      body = postData?.text || ''
+    } else if (reason === 'repost') {
+      title = `${author} reposted your post`
+      body = postData?.text || ''
+
+      // TODO: handle images
+      // let image
+      // if (
+      //   AppBskyEmbedImages.isView(
+      //     notification.additionalPost?.thread?.post.embed,
+      //   ) &&
+      //   notification.additionalPost?.thread?.post.embed.images[0]?.thumb
+      // ) {
+      //   image = notification.additionalPost.thread.post.embed.images[0].thumb
+      // }
+    }
+
+    if (title === '' && body === '') {
+      throw new Error('Failed to get title and body')
```

Comment:
it's actually totally fine to throw in here, if we have this on the background queue & are catching errors

that being said, these errors aren't gonna give us a ton of info & it might be nice to stick some additional fields on them (like the record it had a hard time processing for instance)

alternately, we can not throw & just log then return

Reply:
maybe a custom error class that allows you to stick some additional info on

Reply:
I think logging them is probably fine as they are not critical.